ENH: implement `at`

crusaderky opened the pull request:
Implement a new `at(x, idx)` or `at(x)[idx]` function, mocking the syntax of [JAX's omonymous method](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html) .
This is propaedeutic to JAX support in libraries that support the Array API, e.g. scipy.

Moved from https://github.com/data-apis/array-api-compat/pull/205

# Blockers
- https://github.com/data-apis/array-api-compat/pull/205
- https://github.com/data-apis/array-api-compat/pull/207
- https://github.com/data-apis/array-api-compat/pull/211
# Blocks
- https://github.com/scipy/scipy/pull/22070

crusaderky commented on pyproject.toml:
```diff
@@ -26,7 +26,11 @@ classifiers = [
 ]
 dynamic = ["version"]
-dependencies = ["array-api-compat>=1.1.1"]
+# dependencies = [ "array-api-compat" >= 1.1.1 ]  # DNM
+dependencies = [
+    # DNM main plus #211, #205, #207
+    "array-api-compat @ git+https://github.com/crusaderky/array-api-compat.git@9761bf2c2790416e3c117ed9f64407525b4a3498"
+]
```
This doesn't work and I could use the help of someone more skilled in pixi-fu.
I want to test this PR (locally and in CI) against my latest changes to array-api-compat, before they are pushed into a new release.
The updated pixi.lock looks correct (see above). However, when I run `pixi run -e tests tests` I get
```
E   ImportError: cannot import name 'is_writeable_array' from 'array_api_compat' (/home/crusaderky/github/array-api-extra/.pixi/envs/tests/lib/python3.13/site-packages/array_api_compat/__init__.py)
```
`is_writeable_array` is exactly the function that has been added in the git branch temporarily referenced above.
if I look at `.pixi/envs/tests`, however, I can see that pixi installed the latest released version (1.9.1) instead.

lucascolley replied:
https://pixi.sh/latest/reference/pixi_manifest/#full-specification might help you, otherwise the pixi devs are super helpful on the prefix-dev discord

crusaderky replied:
It was a bug https://github.com/prefix-dev/pixi/issues/2677

crusaderky commented on src/array_api_extra/_funcs.py:
```diff
@@ -546,3 +552,275 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+            raise ValueError(f"Invalid value for copy: {copy!r}")
+
+        if copy:
+            try:
+                at_ = x.at
+            except AttributeError:
```
The alternative to this duck-type pattern is to import `is_jax_array` from array-api-compat. @lucascolley please advise on what you'd prefer.

crusaderky commented on tests/test_at.py:
```diff
@@ -0,0 +1,174 @@
+
+from array_api_extra import at
+
+from ._helpers import all_libraries, import_
```
WIP still need to refactor this.

@lucascolley this test module necessarily needs to add all libraries (numpy, jax, dask, torch, sparse, cupy) to CI in order to test the functionality, whereas the rest of the test suite exclusively tests vs. array-api-strict.
Should I add these libraries to ci.yml or should I leave it alone and create a brand new github workflow?

lucascolley replied:
Seems fine to add to the existing workflow for now. You can create a separate pixi `ci-` env with all the libraries

lucascolley commented on tests/test_at.py:
```diff
@@ -0,0 +1,174 @@
+from ._helpers import all_libraries, import_
+
+
+def assert_array_equal(a, b):
+    if is_pydata_sparse_array(a):
+        a = a.todense()
+    elif is_cupy_array(a):
+        a = a.get()
+    elif is_dask_array(a):
+        a = a.compute()
+    np.testing.assert_array_equal(a, b)
```
See also gh-17

lucascolley commented on pyproject.toml:
```diff
@@ -227,6 +235,9 @@ ignore = [
     "PLR09",   # Too many <...>
     "PLR2004", # Magic value used in comparison
     "ISC001",  # Conflicts with formatter
+    "EM101",   # raw-string-in-exception
```
I would rather fix these, the justification given on the rule page seems reasonable to me: https://docs.astral.sh/ruff/rules/raw-string-in-exception/

crusaderky commented:
- [x] Implementation
- [x] Tests pass locally
- [ ] Make linters happy
- [ ] CI
- [ ] Review docs

crusaderky commented:
Rebased on top of #58

crusaderky commented:
Any idea how to fix this? Looks like the ubuntu_latest VM has an obsolete driver

> ERROR cupy_backends.cuda.api.runtime.CUDARuntimeError: cudaErrorInsufficientDriver: CUDA driver version is insufficient for CUDA runtime version

crusaderky commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,348 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+        self._check_args(copy=copy)
+        x = self.x
+
+        if copy is False:
+            if is_array_api_obj(self.idx):
+                # Boolean index. Note that the array API spec
+                # https://data-apis.org/array-api/latest/API_specification/indexing.html
+                # does not allow for list, tuple, and tuples of slices plus one or more
+                # one-dimensional array indices, although many backends support them.
+                # So this check will encounter a lot of false negatives in real life,
+                # which can be caught by testing the user code vs. array-api-strict.
+                msg = "get() with an array index always returns a copy"
+                raise ValueError(msg)
+
+            # Prevent scalar indices together with copy=False.
+            # Even if some backends may return a scalar view of the original, we chose to be
+            # strict here beceause some other backends, such as numpy, definitely don't.
+            tup_idx = self.idx if isinstance(self.idx, tuple) else (self.idx,)
+            if any(
+                i is not None and i is not Ellipsis and not isinstance(i, slice)
+                for i in tup_idx
+            ):
+                msg = "get() with a scalar index typically returns a copy"
+                raise ValueError(msg)
+
+            # Note: this is not the same list of backends as is_writeable_array()
+            if is_dask_array(x) or is_jax_array(x) or is_pydata_sparse_array(x):
+                msg = f"get() on {array_namespace(x)} arrays always returns a copy"
+                raise ValueError(msg)
```
These checks are very brittle and I'm not comfortable going on with `get()` as it is.

I can see two ways forward:
1. Write a new function in `array-api-compat`, `getitem_returns_view(x: Array, idx: Index) -> bool` that explores all the nooks and crannies of all the possible intersections of arrays and indices, then use it here
2. Remove `get()` altogether. I have a strong suspicion it may not be needed in scipy to begin with.

crusaderky replied:
The only use I can think of out of `get(copy=True)` would be to avoid an unnecessary double copy when idx is a boolean mask

crusaderky replied:
I've removed get(), at least for the time being.

crusaderky commented:
> I don't think we can get GPU CI without paying someone for it, cc @rgommers .

Fixed by swapping `cupy` with `cupy-core`, which is CPU-only.

crusaderky commented:
> I don't think we can get GPU CI without paying someone for it, cc @rgommers .

Can cupy run on a CPU-only host?
I've disabled it in CI for the time being

rgommers commented:
> I don't think we can get GPU CI without paying someone for it, cc @rgommers .

Yep, that's on my radar to push forward this month, on multiple projects. Please feel free to open a new issue and assign it to me. I think we can hook up a shared GPU runner between this project and `array-api-compat`.

> Can cupy run on a CPU-only host?

I don't think so.

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+
+    Note
+    ----
+    `sparse <https://sparse.pydata.org/>`_ is not supported by update methods yet.
```
can you also add a note that read only arrays from any array library not supported by array-api-compat are not supported here?

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+    `jax.numpy.ndarray.at <https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html>`_
+    """
+
+    x: Array
+    idx: Index
```
Currently these attributes render in the docs as

![image](https://github.com/user-attachments/assets/a4b57e80-57ac-4bb7-9ac1-845adf465938)

Could we fix that?

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+    """
+    Update operations for read-only arrays.
+
+    This implements ``jax.numpy.ndarray.at`` for all backends.
```
Can add a clarification along the lines of my comment about read-only arrays from unknown libraries below.

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+    ----------
+    x : array
+        Input array.
+    idx : index, optional
```
you could link to https://data-apis.org/array-api/latest/API_specification/indexing.html in the notes and perhaps reference it here, to explain what an `index` is

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+            Ensure that the inputs are not modified.
+        False
+            Ensure that the update operation writes back to the input.
+            Raise ValueError if a copy cannot be avoided.
```
```suggestion
            Raise ``ValueError`` if a copy cannot be avoided.
```

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+      x = x.copy()
+      x[1] += 2
+
+    Otherwise, they are the same as::
```
```suggestion
    For other known backends, they are the same as::
```

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+
+    Warning
+    -------
+    When you use copy=None, you should always immediately overwrite
```
```suggestion
    When you use ``copy=None``, you should always immediately overwrite
```

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+        y = at(x, 0).set(2, copy=None)
+        z = at(x, 1).set(3, copy=None)
+
+    In the above example, ``x == [0, 0, 0]``, ``y == [2, 0, 0]`` and z == ``[0, 3, 0]``
+    when x is read-only, whereas ``x == y == z == [2, 3, 0]`` when x is writeable!
```
```suggestion
    In the above example, ``x == [0, 0, 0]``, ``y == [2, 0, 0]`` and ``z == [0, 3, 0]``
    when ``x`` is read-only, whereas ``x == y == z == [2, 3, 0]`` when ``x`` is writeable!
```

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -548,3 +561,347 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+    In the above example, ``x == [0, 0, 0]``, ``y == [2, 0, 0]`` and z == ``[0, 3, 0]``
+    when x is read-only, whereas ``x == y == z == [2, 3, 0]`` when x is writeable!
+
+    Warning
+    -------
+    The array API standard does not support integer array indices.
+    The behaviour of update methods when the index is an array of integers
+    is undefined; this is particularly true when the index contains multiple
+    occurrences of the same index, e.g. ``at(x, [0, 0]).set(2)``.
```
for my understanding, can you clarify what this warning means, referencing the standard?

lucascolley reviewed (commented):
overall this looks great to me, thanks! Added some comments.

I'll take another look here before it goes in. In the meantime, hopefully the upstream work goes smoothly, and it would be fantastic if @jakevdp / @rgommers  / anyone else could take a closer look at the implementation!

crusaderky commented on pyproject.toml:
```diff
@@ -225,6 +262,7 @@ ignore = [
     "PLR09",   # Too many <...>
     "PLR2004", # Magic value used in comparison
     "ISC001",  # Conflicts with formatter
+    "PD008",   # Use `.loc` instead of `.at`
```
Everywhere. The linter seems to think that any `Any` object with `.at` is a pandas dataframe.

lucascolley replied:
FWIW I can't reproduce the failures locally - do you have a MWE I could give to the ruff devs?

crusaderky replied:
After I did `rm -rf .pixi pixi.lock`, I cannot reproduce this either anymore. I'm puzzled.

lucascolley replied:
ruff put out a release 1 hour ago so maybe they fixed a bug 🤷‍♂️

crusaderky commented:
I gave a final round of polish. This is ready for final review and approval.
Note the [blocker PRs and DNMs](https://github.com/data-apis/array-api-extra/pull/53#issue-2727108419) before it can be merged though.

crusaderky commented:
> > I don't think we can get GPU CI without paying someone for it, cc @rgommers .
> 
> Yep, that's on my radar to push forward this month, on multiple projects. Please feel free to open a new issue and assign it to me. I think we can hook up a shared GPU runner between this project and `array-api-compat`.
> 
> > Can cupy run on a CPU-only host?
> 
> I don't think so.

https://github.com/data-apis/array-api-extra/issues/60

crusaderky commented on src/array_api_extra/_funcs.py:
```diff
@@ -1,10 +1,22 @@
+from __future__ import annotations
```
Needed due to `__getitem__(self, idx) -> at:`

lucascolley replied:
I think you can remove it if you change that to `__getitem__(self, idx) -> "at":`.

When we drop Python 3.10 we can then change it to `-> Self`.

lucascolley commented:
I resolved the merge conflicts after adding `numpydoc` to pre-commit and made some minor tweaks to the docs.

crusaderky commented:
@lucascolley if you don't mind building against array-api-compat git tip until their next release, this PR is ready for final review and merge

lucascolley commented on pyproject.toml:
```diff
@@ -99,7 +101,9 @@ tests-cov = "pytest -v -ra --cov --cov-report=xml --cov-report=term --durations=
 clean-vendor-compat = "rm -rf vendor_tests/array_api_compat"
 clean-vendor-extra = "rm -rf vendor_tests/array_api_extra"
-copy-vendor-compat = { cmd = "cp -r $(python -c 'import site; print(site.getsitepackages()[0])')/array_api_compat vendor_tests/", depends-on = ["clean-vendor-compat"] }
+copy-vendor-compat = { cmd = "cp -r $(python -c 'import site; print(site.getsitepackages()[0])')/array_api_compat vendor_tests/", depends-on = [
+    "clean-vendor-compat",
+] }
```
this might have been from one of my commits due to an IDE extension, but we should keep this on one line

lucascolley reviewed (commented):
> @lucascolley if you don't mind building against array-api-compat git tip until their next release, this PR is ready for final review and merge

That's fine with me, but might be cleaner to get an array-api-compat release out first unless there are any big blockers.

Let's give @rgommers and @jakevdp time to take a look if they would like to.

Once (or before) this is merged, do you think you could make a PR to my branch at https://github.com/scikit-learn/scikit-learn/pull/30340? I think we will have to transition scikit-learn from using array-api-compat as an optional dependency to vendoring it, but that sounds feasible based on https://github.com/scikit-learn/scikit-learn/issues/30367#issuecomment-2506749233/.

crusaderky commented:
> Once (or before) this is merged, do you think you could make a PR to my branch at [scikit-learn/scikit-learn#30340](https://github.com/scikit-learn/scikit-learn/pull/30340)? I think we will have to transition scikit-learn from using array-api-compat as an optional dependency to vendoring it, but that sounds feasible based on [scikit-learn/scikit-learn#30367 (comment)](https://github.com/scikit-learn/scikit-learn/issues/30367#issuecomment-2506749233).

https://github.com/lucascolley/scikit-learn/pull/2

jakevdp commented on src/array_api_extra/_funcs.py:
```diff
@@ -538,3 +548,314 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+            msg = "Index has already been set"
+            raise ValueError(msg)
+        self._idx = idx
+        return self
```
Maybe it would make sense to return a shallow copy rather than mutating `self`. Otherwise, someone might write something like this and be surprised by the behavior:
```
getter = at(x)
y = getter[0].add(1)
z = getter[1].add(2)
```

crusaderky replied:
Good idea, done

jakevdp commented on src/array_api_extra/_funcs.py:
```diff
@@ -538,3 +548,314 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+        if res is not None:
+            return res
+        assert x is not None
+        x[self._idx] = elwise_op(x[self._idx], y)
```
As currently implemented, the `_iop` path has different semantics for repeated indices between JAX and NumPy:
```python
>>> import numpy as np
>>> x = np.zeros(4)
>>> idx = np.array([1, 2, 2, 3, 3, 3])
>>> x[idx] = x[idx] + 1
>>> x
array([0., 1., 1., 1.])

>>> import jax.numpy as jnp
>>> x = jnp.zeros(4)
>>> idx = jnp.array([1, 2, 2, 3, 3, 3])
>>> x.at[idx].add(1)
Array([0., 1., 2., 3.], dtype=float32)
```
At the very least, the difference should be documented.

crusaderky replied:
it's already documented at lines 612:622

crusaderky commented on pyproject.toml:
```diff
@@ -26,7 +26,7 @@ classifiers = [
 ]
 dynamic = ["version"]
-dependencies = ["array-api-compat>=1.1.1"]
+dependencies = ["array-api-compat>=1.10.0,<2"]
```
@lucascolley Updated to match tool.pixi.dependencies

crusaderky commented on pyproject.toml:
```diff
@@ -62,8 +62,8 @@ channels = ["https://prefix.dev/conda-forge"]
 [tool.pixi.dependencies]
-python = ">=3.10.15,<3.14"
-array-api-compat = ">=1.1.1"
+python = ">=3.10,<3.14"
```
Was there any purpose in specifying the patch version of Python?

crusaderky commented:
@lucascolley ready for merge and release! :partying_face:

lucascolley commented:
good to proceed @rgommers ?

lucascolley commented on src/array_api_extra/_funcs.py:
```diff
@@ -589,3 +599,313 @@ def pad(
+    --------
+    Given either of these equivalent expressions::
+
+      x = at(x)[1].add(2, copy=None)
+      x = at(x, 1).add(2, copy=None)
```
this doesn't have to land in this PR, but I think it would be good to use `import array_api_extra as xpx` ... `x = xpx.at(...)` as that is what is used in the other examples in the docs

crusaderky replied:
Done; let me know if it's what you meant

lucascolley commented:
looks great, thanks! Note to self: we ought to add

```
# sphinx-copybutton configurations
copybutton_prompt_text = r">>> |\.\.\. |\$ |In \[\d*\]: | {2,5}\.{3,}: | {5,8}: "
copybutton_prompt_is_regexp = True
```

to `docs/conf.py`.

crusaderky commented:
Annoyingly, codecov is glitching (again). It's stating in the "Files" view that the tests never run under JAX, but if you follow the [link from Check ci-backends](https://app.codecov.io/github/data-apis/array-api-extra/commit/98141cae091c4cea2cf6df7f4f2eefd2eb235605/blob/src/array_api_extra/_funcs.py) it says that everything's green.

lucascolley commented:
yeah, happy to just ignore that, but also happy if anyone figures out how to fix that

rgommers reviewed (commented):
Thanks @crusaderky and reviewers. This is looking quite good. A couple of comments (disclaimer: it's a little hard to dig through the already-resolved review comments, so I may be saying something that was already discussed there):

(1) The `copy=True` default seems wrong to me. The main point of this feature is that we can replace in-place operations like `x[idx, :] += y` in SciPy et al. with something new to support JAX. That something new should be the idiomatic JAX way, so `x = xpx.at(x)[idx, :].add(y)`. This should come with as little of a performance penalty for NumPy usage as possible - hence the `copy=True` default doesn't work. It's already a significant concession to code readability to have to avoid `+=` & co, and having to also add `copy=None` in all places where performance matters seems too much.

(2) When playing around with this branch, most things work as expected, but this was a little surprising:
```python
>>> x = np.arange(5)
>>> x = xpx.at(x)[[3, 4]].divide(1.5)
>>> x.dtype
dtype('int64')
>>> x = np.arange(5)
>>> x /= 1.5
Traceback (most recent call last):
  Cell In[32], line 1
    x /= 1.5
UFuncTypeError: Cannot cast ufunc 'divide' output from dtype('float64') to dtype('int64') with casting rule 'same_kind'
```
Is this on purpose or by accident? Raising on these cases may be cleaner if that can be done in a performant way.

(3) Library support is pretty comprehensive already; a few others come to mind. MLX is already supported by the generic path since [it allows in-place syntax](https://ml-explore.github.io/mlx/build/html/usage/indexing.html#in-place-updates) (I think, didn't verify). It also [has array.at](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.array.at.html#mlx.core.array.at) which matches JAX syntax/semantics. `ndonnx` is going to need explicit support in the future as well, since it's immutable like JAX, but that can be left for later since it doesn't seem to have `Array.at` or another alternative here. tl;dr nothing to change in this PR I think.

crusaderky commented on src/array_api_extra/_funcs.py:
```diff
@@ -589,3 +599,318 @@ def pad(
+        Allow for the alternate syntax ``at(x)[start:stop:step]``.
+
+        It looks prettier than ``at(x, slice(start, stop, step))``
+        and feels more intuitive coming from the JAX documentation.
```
I've been writing several prototype PRs on various submodules of Sphinx.
From the experience I gathered there, I must say that `at(x)[idx]` feels more natural and easier on the eyes when there are multiple indices and/or slies, whereas `at(x, idx)` feels nicer and less cluttered in all other cases.

rgommers replied:
Okay, fair enough - seems like an argument to keep both then.

crusaderky replied:
In fact, I've been considered writing aliases for the benefit of readability `set_at(x, idx, value)` etc. but haven't gone for them for the sake of keeping the API surface small.

rgommers replied:
Not adding a third way sounds like a healthy idea at least for now - let's see how we like this in its current form I'd say.

crusaderky commented:
> (1) The copy=True default seems wrong to me.

Changed it to copy=None

> (2) x /= 1.5

TIL that 

- ``x[idx] = x[idx].__add__(value)``
- ``x[idx] = x[idx].__iadd__(value)``

are _almost_ always equivalent!

Fixed.

lucascolley commented on tests/test_at.py:
```diff
@@ -0,0 +1,158 @@
+
+@pytest.mark.parametrize("copy", [True, False])
+@pytest.mark.parametrize("op", ["add", "subtract", "multiply", "divide", "power"])
+def test_iops_incompatible_dtype(op, copy):
```
```suggestion
def test_iops_incompatible_dtype(op: str, copy: bool):
```

lucascolley commented:
shall we get going with a release? :)

rgommers reviewed (approved):
Looks ready to me now, +1 for merge & release if everyone else is happy too.

lucascolley commented:
let's give this a whirl, thanks all! if you can point the SciPy PR to the release when it is out @crusaderky, I can update the sklearn PR
